### PR TITLE
SLOS Tree

### PR DESCRIPTION
**Pull request by ericbrts**

### CODE REVIEW

File: perceval/utils/statevector.py [hunk 1]
```diff
@@ -106,27 +109,28 @@ def add(self, obj, proba: float):
...
     @abstractmethod
-    def sample(self, count: int, non_null: bool = True):
+    def sample(self, count: int, non_null: bool = True) -> list:
         pass
 
 
 class SVDistribution(ProbabilityDistribution):
-    r"""Time-Independent Probabilistic distribution of StateVectors
+    r"""
+    Mixed state represented os a time-independent probabilistic distribution of StateVectors
```
Comment (raksharuia) [inline perceval/utils/statevector.py:133]: typing error - 'represented AS a time-...'

File: perceval/backends/_slos_tree.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        """
+        Use Steiner walk into the tree rather than the naive walk.
+        Steiner's method is more efficient when the number of photons is close to the number of modes.
+        """
+        self._stree.set_steiner(use_it)
+
+    def prob_amplitude(self, output_state: BasicState) -> complex:
+        istate = self._input_state
+        all_pa = self._stree.all_prob_ampli(istate)
+        return all_pa[xq.FSArray(istate.m, istate.n).find(output_state)]
```
Comment (Aubaert) [inline perceval/backends/_slos_tree.py:57]: Do we need to create this FSArray everytime?

File: perceval/backends/_slos_tree.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    def prob_distribution(self) -> BSDistribution:
+        istate = self._input_state
+        all_probs = self._stree.all_prob(istate)
+
+        if self._mask is not None:  # Utterly non-optimized. Mask management should be added in the computation
+            fsa = xq.FSArray(istate.m, istate.n)
```
Comment (Aubaert) [inline perceval/backends/_slos_tree.py:64]: Same here

File: perceval/backends/_slos_tree.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            self.set_input_state(input_state)
+        else:
+            input_state = self._input_state
+        return self._stree.all_prob(input_state)
+
+    def evolve(self) -> StateVector:
+        istate = self._input_state
+        all_pa = self._stree.all_prob_ampli(istate)
+        res = StateVector()
+        for output_state, pa in zip(xq.FSArray(istate.m, istate.n), all_pa):
```
Comment (Aubaert) [inline perceval/backends/_slos_tree.py:87]: Same here

File: perceval/backends/_slos_tree.py (hunk 2 above)
Comment (ericbrts) [reply perceval/backends/_slos_tree.py]: Fixed

**Pull request merged**